### PR TITLE
Crank 2: Let's Get Crankier

### PR DESCRIPTION
**Pull request by anurse**
[Crank](https://github.com/SignalR/DCrank) returns! The bulk of this is a port of that code to .NET Core and ASP.NET Core, but see below for some more info.

This PR adds a new .NET Core Console app called "Crankier" to the benchmarks folder. There are three "commands" in this app:

* `local` - Runs a self-contained Agent
* `agent` - Runs an Agent that connects to a controller (not yet implemented)
* `worker` - Runs a single Worker node (launched automatically by the Agent)

The Agent spins up a series of Worker processes, communicating via stdio. Each Worker start a number of connections and send status updates to the Agent. Once all connections across all Workers have been established, the Agent instructs the Workers to send a payload to the server at a regular interval, for a specified duration. Finally, when the duration elapses, the Workers stop all their connections.

It's configured to work properly against the BenchmarkServer EchoHub but any Hub with the necessary method will work.

Other notable changes:
* It connects at the Hub level, because a) that was easier and b) it means we get the server sending us heartbeats (driving traffic that way)
* It doesn't use our coding style. I'll try to do a formatting pass this weekend.

### CODE REVIEW

File: benchmarks/Crankier/WorkerReceiver.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    }
+                }
+            });
+        }
+
+        public void Stop()
+        {
+            _receiveMessageCts.Cancel();
+        }
+
```
Comment (JamesNK) [inline benchmarks/Crankier/WorkerReceiver.cs]: nit

File: benchmarks/Crankier/Worker.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+        private async Task SendStatusUpdate(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                int connectedCount = 0;
+                int connectingCount = 0;
+                int disconnectedCount = 0;
+                int reconnectingCount = 0;
+                int faultedCount = 0;
```
Comment (JamesNK) [inline benchmarks/Crankier/Worker.cs]: vaarrrr

File: benchmarks/Crankier/Worker.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            await _agent.Pong(_processId, value);
+            Log("Worker sent pong command with value {0}.", value);
+        }
+
+        public async Task Connect(string targetAddress, HttpTransportType transportType, int numberOfConnections)
+        {
+            Log("Worker received connect command with target address {0} and number of connections {1}", targetAddress, numberOfConnections);
+
+            _targetConnectionCount += numberOfConnections;
+            for (int count = 0; count < numberOfConnections; count++)
```
Comment (JamesNK) [inline benchmarks/Crankier/Worker.cs]: var

**Comment by JamesNK** [comment]
And copyright header everywhere.

File: benchmarks/BenchmarkServer/Hubs/EchoHub.cs [hunk 4]
```diff
@@ -39,5 +40,11 @@ public Task EchoAll(DateTime time)
         {
             return Clients.All.SendAsync("send", time);
         }
+
+        public Task SendPayload(string payload)
```
[the remark this answers is not in the document]
Comment (anurse) [reply benchmarks/BenchmarkServer/Hubs/EchoHub.cs]: Yeah, all part of tidying up the old code :). Will do that, and `var` all the things.

**Comment by anurse** [comment]
:up: :date: Got copyright headers added and style mostly in sync.

**Pull request merged**